### PR TITLE
[Feature] Support iSAID aerial dataset.

### PR DESCRIPTION
**Pull request by Waterman0524**
Enhance the dataset diversity for MMSegmentation in Remote Sensing.

## Update (2022.01.26)

<div class="okr-block-clipboard" data-okr="%7B%22okrDelta%22%3A%5B%7B%22lineType%22%3A%22unsupport%22%2C%22lineOptions%22%3A%7B%7D%2C%22lineContent%22%3A%5B%5D%7D%5D%2C%22businessKey%22%3A%22lark-doc%22%7D"></div><div style="white-space: pre;" data-zone-id="0" data-line-index="0">

Config names | mIoU (validation) | mIoU (ms + flip) | GPU memory (MB) | FPS
-- | -- | -- | -- | --
pspnet_r18-d8_4x4_896x896_80k_isaid | 60.22 | 61.25 | 4626 | 24.81
pspnet_r50-d8_4x4_896x896_80k_isaid | 65.36 | 66.48 | 16981 | 8.42
deeplabv3plus_r18-d8_4x4_896x896_80k_isaid | 61.35 | 62.61 | 6335 | 26.91
deeplabv3plus_r50-d8_4x4_896x896_80k_isaid | 67.06 | 68.02 | 21967 | 8.88
fcn_hr18s_4x4_896x896_80k_isaid | 62.3 | 62.97 | 5068 | 13.84
fcn_hr18_4x4_896x896_80k_isaid | 65.06 | 65.60 | 8501 | 7.71
fcn_hr48_4x4_896x896_80k_isaid | 67.8 | 68.53 | 17293 | 7.34

</div>

## TODO

- [x] Get great numerical results
- [x] Fix unit test coverage rate
- [x] Upload models & logs.

### CODE REVIEW

File: 博士/dataset_prepare.md [hunk 1]
```diff
@@ -286,3 +286,33 @@ python tools/convert_datasets/loveda.py /path/to/loveDA
 Using trained model to predict test set of LoveDA and submit it to server can be found [here](https://github.com/open-mmlab/mmsegmentation/blob/master/docs/inference.md).
 
 More details about LoveDA can be found [here](https://github.com/Junjue-Wang/LoveDA).
+
```
Comment (Junjun2016) [inline 博士/dataset_prepare.md:289]: Should not rename this file.

File: 博士/dataset_prepare.md [hunk 2]
```diff
@@ -286,3 +286,33 @@ python tools/convert_datasets/loveda.py /path/to/loveDA
...
+ --val
+   --images
+     --part1.zip
+   --Semantic_masks
+     --images.zip
+ --test 
+   --images
+     --part1.zip
+     --part2.zip
+
```
Comment (Junjun2016) [inline 博士/dataset_prepare.md:314]:
```
|   │   |   ├── val
|   │   |   └── val_ref
|   ├── NighttimeDrivingTest
|   |   ├── gtCoarse_daytime_trainvaltest
|   |   │   └── test
|   |   │       └── night
|   |   └── leftImg8bit
|   |   |   └── test
|   |   |       └── night
│   ├── loveDA
│   │   ├── img_dir
│   │   │   ├── train
│   │   │   ├── val
│   │   │   ├── test
│   │   ├── ann_dir
│   │   │   ├── train
│   │   │   ├── val
```

Use this format.

File: 博士/dataset_prepare.md (hunk 1 above)
Comment (Waterman0524) [reply 博士/dataset_prepare.md:297]: Sorry, I don't know why this happened

File: 博士/dataset_prepare.md (hunk 2 above)
Comment (Waterman0524) [reply 博士/dataset_prepare.md:323]: OK

**Comment by MengzhangLI** [comment]
Hi, @Waterman0524  . Could you grant authorization on me about your [forked MMSegmentation](https://github.com/Waterman0524/mmsegmentation/tree/mmsegmentation-2) followed [here](https://docs.github.com/cn/account-and-profile/setting-up-and-managing-your-github-user-account/managing-access-to-your-personal-repositories/inviting-collaborators-to-a-personal-repository)? Thus I could push my modifications on your branch. 

image


Right now there is lint error by our CI in Github. I could use my local `pre-commit` command to swipe them.

Best,

**Comment by Junjun2016** [comment]
Add class name and palette to `mmseg/core/evaluation/class_names.py`.

File: docs/zh_cn/dataset_prepare.md [hunk 3]
```diff
@@ -266,3 +278,37 @@ python tools/convert_datasets/vaihingen.py /path/to/vaihingen
...
+│   │   │   ├── images
+│   │   │   │   ├── part1.zip
+│   │   │   │   ├── part2.zip
+```
+
+```shell
+python tools/convert_datasets/isaid.py /path/to/iSAID
+```
+
+使用我们默认的配置 (`clip_size` =512, `stride_size`=256)， 将生成 33978 张图片的训练集和 11644 张图片的验证集。
```
Comment (Junjun2016) [inline docs/zh_cn/dataset_prepare.md:314]:
```suggestion
使用我们默认的配置 (`clip_size`=512, `stride_size`=256)， 将生成 33978 张图片的训练集和 11644 张图片的验证集。
```

File: mmseg/datasets/iSAID.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import mmcv
+from mmcv.utils import print_log
+
+from ..utils import get_root_logger
+from .builder import DATASETS
+from .custom import CustomDataset
+
+
+@DATASETS.register_module()
+class ISAIDDataset(CustomDataset):
```
Comment (Junjun2016) [inline mmseg/datasets/iSAID.py:12]: Use `iSAIDDataset` instead?

File: configs/_base_/datasets/iSAID.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
+# dataset settings
+dataset_type = 'ISAIDDataset'
+data_root = 'data/iSAID'
+
+img_norm_cfg = dict(
+    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
+
+crop_size = (896, 896)
```
Comment (Junjun2016) [inline configs/_base_/datasets/iSAID.py:8]: 896 or 512?

File: mmseg/datasets/iSAID.py (hunk 4 above)
Comment (MengzhangLI) [reply mmseg/datasets/iSAID.py]: OK

File: configs/_base_/datasets/iSAID.py (hunk 5 above)
Comment (MengzhangLI) [reply configs/_base_/datasets/iSAID.py:8]: It should be 896. Sorry for misleading number 512 in config names and docs.

Comment (Junjun2016) [reply configs/_base_/datasets/iSAID.py:8]: Have experiments with crop size 512.

File: docs/en/dataset_prepare.md [hunk 6]
```diff
@@ -123,6 +123,18 @@ mmsegmentation
...
+│   │   │   ├── train
+│   │   │   ├── val
+│   │   ├── ann_dir
+│   │   │   ├── train
+│   │   │   ├── val
+│   ├── iSAID
+│   │   ├── img_dir
+│   │   │   ├── train
+│   │   │   ├── val
+│   │   │   ├── test
```
Comment (RockeyCoss) [inline docs/en/dataset_prepare.md:137]: forget to write ann_dir

File: mmseg/datasets/iSAID.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        Returns:
+            list[dict]: All image info of dataset.
+        """
+
+        img_infos = []
+        if split is not None:
+            with open(split) as f:
+                for line in f:
+                    name = line.strip()
+                    img_name = name
```
Comment (RockeyCoss) [inline mmseg/datasets/iSAID.py:61]: Is it necessary to rename `name` to `img_name` ?

**Review by MengzhangLI — APPROVED** [review]

File: docs/en/dataset_prepare.md [hunk 8]
```diff
@@ -325,3 +340,37 @@ python tools/convert_datasets/vaihingen.py /path/to/vaihingen
...
+
+### iSAID
+The data images could be download from [DOTA-v1.0](https://captain-whu.github.io/DOTA/dataset.html) (train/val/test)
+
+The data annotations could be download from [iSAID](https://captain-whu.github.io/iSAID/dataset.html) (train/val)
+
+The dataset is a Large-scale Dataset for Instance Segmentation (also have segmantic segmentation) in Aerial Images.
+
+You may need to follow the following structure for dataset preparation after downloading iSAID dataset.
+```
```
Comment (MeowZheng) [inline docs/en/dataset_prepare.md:352]: Add a blank line between ```

File: docs/zh_cn/dataset_prepare.md [hunk 9]
```diff
@@ -265,4 +280,38 @@ python tools/convert_datasets/potsdam.py /path/to/potsdam
...
+
+### iSAID
+iSAID 数据集(训练集/验证集/测试集)的图像可以从 [DOTA-v1.0](https://captain-whu.github.io/DOTA/dataset.html) 下载.
+
+iSAID 数据集(训练集/验证集)的注释可以从 [iSAID](https://captain-whu.github.io/iSAID/dataset.html) 下载.
+
+该数据集是一个大规模的实例分割(也可以用于语义分割)的遥感数据集.
+
+下载后，在数据集转换前，您需要将数据集文件夹调整成如下格式.
+```
```
Comment (MeowZheng) [inline docs/zh_cn/dataset_prepare.md:293]: add a blank line before ```

File: docs/en/dataset_prepare.md (hunk 8 above)
Comment (MengzhangLI) [reply docs/en/dataset_prepare.md:353]: Added in L352.

File: configs/_base_/datasets/iSAID.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
+# dataset settings
+dataset_type = 'iSAIDDataset'
+data_root = 'data/iSAID'
+
+img_norm_cfg = dict(
+    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
+
+crop_size = (896, 896)
```
Comment (MeowZheng) [inline configs/_base_/datasets/iSAID.py:8]: add comments to present the reference of this setting

File: configs/deeplabv3plus/README.md [hunk 11]
```diff
@@ -114,6 +114,13 @@ Spatial pyramid pooling module or encode-decoder structure are used in deep neur
...
+| DeepLabV3+ | R-18-D8  | 896x896   |   80000 | 6.19      | 24.81             | 61.35 |         62.61 | [config](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/deeplabv3plus/deeplabv3plus_r18-d8_4x4_896x896_80k_isaid.py)  | [model](https://download.openmmlab.com/mmsegmentation/v0.5/deeplabv3plus/deeplabv3plus_r18-d8_4x4_896x896_80k_isaid/deeplabv3plus_r18-d8_4x4_896x896_80k_isaid_20220110_180526-7059991d.pth) &#124; [log](https://download.openmmlab.com/mmsegmentation/v0.5/deeplabv3plus/deeplabv3plus_r18-d8_4x4_896x896_80k_isaid/deeplabv3plus_r18-d8_4x4_896x896_80k_isaid_20220110_180526.log.json)     |
```
Comment (MeowZheng) [inline configs/deeplabv3plus/README.md:121]: add footnote for this setting

File: mmseg/datasets/iSAID.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    def __init__(self, **kwargs):
+        super(iSAIDDataset, self).__init__(
+            img_suffix='.png',
+            seg_map_suffix='.png',
+            ignore_index=255,
+            **kwargs)
+        assert osp.exists(self.img_dir)
+
+    def load_annotations(self, img_dir, img_suffix, ann_dir, seg_map_suffix,
+                         split):
```
Comment (MeowZheng) [inline mmseg/datasets/iSAID.py:40]: split default to None, based on the docstring

File: mmseg/datasets/iSAID.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        """Load annotation from directory.
+
+        Args:
+            img_dir (str): Path to image directory
+            img_suffix (str): Suffix of images.
+            ann_dir (str|None): Path to annotation directory.
+            seg_map_suffix (str|None): Suffix of segmentation maps.
```
Comment (MeowZheng) [inline mmseg/datasets/iSAID.py:47]: if arguments default value is `None`, the type can be `(xx | None)`, if not, the type is  (xx)

**Review by MeowZheng — APPROVED** [review]
revise mmseg/datasets file and config file  iSAID.py to isaid.py

**Comment by OpenMMLab-Assistant003** [comment]
Hi @Waterman0524！First of all, we want to express our gratitude for your significant PR in the MMSegmentation project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤ @Waterman0524

**Pull request merged**